@here/harp-mapview/lib/MapViewAtmosphere.ts

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Theme } from "@here/harp-datasource-protocol";
+import { EarthConstants, Projection, ProjectionType } from "@here/harp-geoutils";
+
+import { GroundAtmosphereMaterial, SkyAtmosphereMaterial } from "@here/harp-materials";
+import { MapAnchor, MapView } from "./MapView";
+
+import { assert } from "@here/harp-utils";
+import * as THREE from "three";
+
+/**
+ * Atmosphere effect variants.
+ */
+enum AtmosphereVariant {
+    Ground = 0x1,
+    Sky = 0x2,
+    SkyAndGround = 0x3
+}
+
+/**
+ * Atmosphere shader variants.
+ */
+enum AtmosphereShadingVariant {
+    ScatteringShader,
+    SimpleColor,
+    Wireframe
+}
+
+/**
+ * Lists light modes.
+ */
+export enum AtmosphereLightMode {
+    LightOverhead = 0,
+    LightDynamic = 1
+}
+
+/**
+ * Class that provides [[MapView]]'s atmospheric scattering effect.
+ */
+export class MapViewAtmosphere {
+    /**
+     * User data name attribute assigned to created mesh.
+     */
+    static SkyAtmosphereUserName: string = "SkyAtmosphere";
+    /**
+     * User data name attribute assigned to created mesh.
+     */
+    static GroundAtmosphereUserName: string = "GroundAtmosphere";
+
+    /**
+     * Check if scene and map view has already atmosphere effect added.
+     * @param where [[MapView]] or [[THREE.Scene]] instance.
+     */
+    static isPresent(where: MapView | THREE.Scene): boolean {
+        const scene = where instanceof MapView ? where.scene : where;
+        if (scene.getObjectByName(MapViewAtmosphere.SkyAtmosphereUserName)) {
+            return true;
+        } else if (scene.getObjectByName(MapViewAtmosphere.GroundAtmosphereUserName)) {
+            return true;
+        }
+        return false;
+    }
+
+    private m_enabled: boolean = true;
+    private m_skyGeometry?: THREE.BufferGeometry;
+    private m_skyMaterial?: THREE.Material;
+    private m_skyMesh?: THREE.Mesh;
+    private m_groundGeometry?: THREE.BufferGeometry;
+    private m_groundMaterial?: THREE.Material;
+    private m_groundMesh?: THREE.Mesh;
+    // TODO: Support for Theme definition should be added.
+    //private m_cachedTheme: Theme = { styles: {} };
+
+    private readonly m_lightDirection = new THREE.Vector3(0.0, 1.0, 0.0);
+
+    /**
+     * Creates and adds `Atmosphere` effects.
+     *
+     * @note Currently works only with globe projection.
+     *
+     * @param m_mapView [[MapView]] instance where the effect will be added.
+     * @param m_atmosphere Atmosphere configuration parameters.
+     */
+    constructor(
+        private m_mapView: MapView,
```

Comment:
Why do we need full MapView? I think we only need scene, projection and camera.
Passing MapView everywhere is like cancer in our codebase and at least for new code we should stop it.

Reply:
Of course I know about that... and wasn't very happy to add it here, but unfortunately I need the access to MapView.anchors and MapView.scene. What I could do anyway is to replace map view with anchors list and scene reference, that should be enough to configure this component... What do you think?

Reply:
@ninok unfortunately when I implemented full integration with MapView component, so there is no need for special ClipPlanesEvaluator, I am even more constraint to this reference, cause I need to access current view ranges.... See my comments below.

Reply:
@kkostecki Still you could pass whatever you need e.g. in the update method.
I'm also wondering if we should really use the map anchor api for the atmosphere as well. We can do the world position update also in the atmosphere update.
Would be weird for the user to encounter the atmosphere when iterating through the map anchors. You would only expect your own objects.

Reply:
@ninok regarding your last comment - of course we may not use anchors, but then we need to re-implement (duplicate) entire logic of objects handling - related to map position. You probably remember that on globe camera is in origin, but entire globe is transformed. Similar (but different transforms) situation we will have for planar projection (if ever supported with the same component).
Another question, how would you like to `update` MapViewAtmosphere` component? Do you want to create MapView - > Atmosphere dependency or create some generic container for "other" objects handled via MapView? Please clarify.

Reply:
Not a bad idea... The question is if we wont to grow this PR with all this changes? It would of course require MapViewFog refactoring.